### PR TITLE
feat: Full Endpoint Resolution from EndpointContext

### PR DESCRIPTION
**Pull request by lqiu96**
Features:
- Introduces the concept of Universe Domain
- Endpoint + Universe Domain resolution is determined solely by the EndpointContext
- Support and compatibility for GDC-H, DirectPath, and mTLS

### CODE REVIEW

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java [hunk 1]
```diff
@@ -157,15 +160,29 @@ public static ClientContext create(StubSettings settings) throws IOException {
...
-    if (credentials instanceof GdchCredentials) {
+    if (usingGDCH) {
       // We recompute the GdchCredentials with the audience
       String audienceString;
       if (!Strings.isNullOrEmpty(settingsGdchApiAudience)) {
         audienceString = settingsGdchApiAudience;
-      } else if (!Strings.isNullOrEmpty(settings.getEndpoint())) {
-        audienceString = settings.getEndpoint();
+      } else if (!Strings.isNullOrEmpty(endpoint)) {
+        audienceString = endpoint;
```
Comment (lqiu96) [inline gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:185]: Confirm this behavior is correct

Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:185]: The GDCH audienceString's endpoint should match with endpoint set inside the TransportChannelProvider. Use the EndpointContext's resolved endpoint for both.

Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:185]: The else block is kept as the custom endpoint could be set as "" (empty string). `URI.create("")` does not throw an exception, so it is caught in the else block.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 2]
```diff
@@ -69,28 +77,76 @@ public abstract class EndpointContext {
...
+    // Universe Domain defaults to the GDU if it's not provided by the user.
+    resolvedUniverseDomain = universeDomain() != null ? universeDomain() : GOOGLE_DEFAULT_UNIVERSE;
+
+    if (Strings.isNullOrEmpty(customEndpoint)) {
+      customEndpoint = buildEndpointTemplate(serviceName(), resolvedUniverseDomain);
+      resolvedEndpoint =
+          mtlsEndpointResolver(
+              customEndpoint, mtlsEndpoint(), switchToMtlsEndpointAllowed(), mtlsProvider);
+      // Check if mTLS is configured with non-GDU
+      if (resolvedEndpoint.equals(mtlsEndpoint())
```
Comment (blakeli0) [inline gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:121]: This validation works for now but there some potential concerns. Ideally we should not set `resolvedEndpoint` until the validation is passed. Here we set the resolved endpoints first, then use the set endpoint for validation. In case there are some changes to the validation logic, e.g. log warnings instead of throwing exception, it would be error prone.

A fix would be to either create a temp variable to hold the value and set it as the last step, or preferably we change this method to return a resolved endpoint, and set it separately. So that it separates the concerns of calculating and setting resolved endpoints.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 3]
```diff
@@ -69,28 +77,76 @@ public abstract class EndpointContext {
...
+        throw new IllegalArgumentException(
+            "mTLS is not supported in any universe other than googleapis.com");
+      }
+    } else {
+      resolvedEndpoint = customEndpoint;
+    }
+  }
+
+  // GDC-H has no concept of Universe Domain. Do not set the resolvedUniverseDomain value
+  private void determineGDCHEndpoint(String customEndpoint) {
```
Comment (blakeli0) [inline gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:132]: We can change this method to return a resolved GDCH endpoint as well, instead of resolving and setting in the same place, if you plan to change the `determineEndpoint()` to return a resolved endpoint.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java [hunk 4]
```diff
@@ -104,6 +104,9 @@ public abstract class ClientContext {
   @Nullable
   abstract String getServiceName();
 
+  @Nullable
+  public abstract String getUniverseDomain();
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:108]: This is for the for logic that connects ClientContext <-> StubSettings. I believe it was added a while back for reasons and this getter is (ideally) only for StubSettings usage.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 5]
```diff
@@ -33,13 +33,18 @@
...
 import javax.annotation.Nullable;
 
-/** Contains the fields required to resolve the endpoint */
+/** Contains the fields required to resolve the endpoint and Universe Domain */
 @InternalApi
 @AutoValue
 public abstract class EndpointContext {
+  // Default to port 443 for HTTPS. Using HTTP requires explicitly setting the endpoint
+  private static final String ENDPOINT_TEMPLATE = "SERVICE_NAME.UNIVERSE_DOMAIN:443";
+  static final String GOOGLE_DEFAULT_UNIVERSE = "googleapis.com";
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:46]: Yeah. We can migrate to that constant once it's released.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 6]
```diff
@@ -69,28 +77,76 @@ public abstract class EndpointContext {
...
+    } else if (customEndpoint == null) {
+      resolvedEndpoint = buildEndpointTemplate(serviceName(), GOOGLE_DEFAULT_UNIVERSE);
+    } else {
+      resolvedEndpoint = customEndpoint;
+    }
+  }
+
+  // Construct the endpoint with the template
+  private String buildEndpointTemplate(String serviceName, String resolvedUniverseDomain) {
+    return ENDPOINT_TEMPLATE
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:145]: Sure, sounds good. I'll change.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 7]
```diff
@@ -69,28 +77,76 @@ public abstract class EndpointContext {
...
+    if (usingGDCH()) {
+      determineGDCHEndpoint(customEndpoint);
+      return;
+    }
+    // Check for "" (empty string)
+    if (universeDomain() != null && universeDomain().isEmpty()) {
+      throw new IllegalArgumentException("The universe domain value cannot be empty.");
+    }
+    // Universe Domain defaults to the GDU if it's not provided by the user.
+    resolvedUniverseDomain = universeDomain() != null ? universeDomain() : GOOGLE_DEFAULT_UNIVERSE;
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:113]: Sounds good, will do.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 2 above)
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:121]: Makes sense. I'll update.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 3 above)
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:132]: Sure, will update.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java]: > Can we default the universe domain to GDU instead of introducing a new field resolvedUniverseDomain?

Do we still need `resolvedUniverseDomain`?

[the remark this answers is not in the document]
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java]: I planned to remove it and use GDU by default, but I think there is a small edge case. Found it when running the IT GDCH tests, specifically regarding: https://github.com/googleapis/sdk-platform-java/blob/3d885bab4e3fd71f682e3d477c4cb07337629681/showcase/gapic-showcase/src/test/java/com/google/showcase/v1beta1/it/ITGdch.java#L189-L192.

L190's ClientContext.create(...) creates an EndpointContext. Nothing is passed in via the universeDomain getter as it's a GDC-H flow, but the universe domain would be set to GDU by default.

L192's createStub() will create the Stub and it takes the params from the ClientContext (which has the GDU as the universe domain) and it will fail the GDC-H logic block.

I'm not sure the way that the IT GDC-H tests are set is makes sense (or if I'm missing something), specifically regarding explicitly creating the ClientSettings -> ClientContext -> StubSettings steps. But I think it shows that doing this is possible by the user.

I learned towards keeping the resolvedUniverseDomain so that the user can differentiate between what the user configuration and what the resolved universe domain is (null, GDU, user configured). This would possibly be done in a future enhancement as a getter.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 8]
```diff
@@ -149,12 +118,124 @@ public abstract static class Builder {
...
+    abstract boolean usingGDCH();
+
+    abstract String resolvedUniverseDomain();
+
     abstract EndpointContext autoBuild();
 
+    private String determineUniverseDomain() {
+      // Do not set the universe domain for GDC-H
+      if (usingGDCH()) {
+        return null;
```
Comment (blakeli0) [inline gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:150]: Can we return GDU here instead of null? And make `resolvedUniverseDomain` not nullable? I see that `resolvedUniverseDomain` is used multiple places later, it is probably good now but may easily introduce null pointer exceptions later.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 9]
```diff
@@ -149,12 +118,124 @@ public abstract static class Builder {
...
+      return endpoint;
+    }
+
+    // GDC-H has no concept of Universe Domain. Do not set the resolvedUniverseDomain value
+    private String determineGDCHEndpoint(String customEndpoint) {
+      if (universeDomain() != null) {
+        throw new IllegalArgumentException(
+            "Universe domain configuration is incompatible with GDC-H");
+      } else if (customEndpoint == null) {
+        return buildEndpointTemplate(serviceName(), GOOGLE_DEFAULT_UNIVERSE);
```
Comment (blakeli0) [inline gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:199]: Can we use `resolvedUniverseDomain` if we default it to GDU?

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java]: > I learned towards keeping the resolvedUniverseDomain so that the user can differentiate between what the user configuration and what the resolved universe domain is (null, GDU, user configured)

Makes sense. Please see my new comments regarding `resolvedUniverseDomain`, I think it's better to make it not nullable, but maybe there is a edge case preventing us from doing so.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (blakeli0) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/ClientContext.java:108]: Yeah, for every new `ClientSettings`, it seems that we have to duplicate it in `StubSettings` as well.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 10]
```diff
@@ -69,58 +75,19 @@ public abstract class EndpointContext {
...
-  public abstract Builder toBuilder();
+  public abstract boolean usingGDCH();
 
-  private String resolvedEndpoint;
-
-  public static Builder newBuilder() {
-    return new AutoValue_EndpointContext.Builder().setSwitchToMtlsEndpointAllowed(false);
-  }
+  @Nullable
+  public abstract String resolvedUniverseDomain();
```
Comment (blakeli0) [inline gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:81]: Do we expect `resolvedUniverseDomain` to be used by anything outside of this class? If not, I think we can make the this method and the setter in the builder package private (I don't think we can make it completed private though due to AutoValue limitations).

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 8 above)
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:150]: The resolvedUniverseDomain is what is stored in the ClientContext and StubSettings (not the user configured universe domain). Following the GDC-H case above, it's possible to manually construct the StubSettings and ClientContext and each would construct the EndpointContext. If we set the resolvedUniverseDomain to the GDU by default, then it would fail every case as the Universe Domain can never be set for GDC-H (must be null).

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 10 above)
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java]: Actually on second thought, I think you're right. I've made it package private. I originally used it as the value I set in the ClientContext, but I can just follow what we are doing with endpoint and pass the user set values for that. This way the resolvedUniverseDomain can be non-null (default to GDU).

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java (hunk 9 above)
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java]: Updated. Thanks!

**Review by blakeli0 — APPROVED** [review]
LGTM. Please add Javadocs before merging.

File: gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java [hunk 11]
```diff
@@ -48,6 +51,9 @@ public abstract class EndpointContext {
   @Nullable
   public abstract String serviceName();
 
+  @Nullable
```
[the remark this answers is not in the document]
Comment (lqiu96) [reply gax-java/gax/src/main/java/com/google/api/gax/rpc/EndpointContext.java:60]: Done. Thanks
